Feature/placed pages

danielebogo opened the pull request:
This PR adds the Properly Placed Pages feature.

![ppp](https://user-images.githubusercontent.com/912252/45882930-89917700-bda7-11e8-9c18-5e43e32aeae5.jpg)

**Pod:**
WordPressKi-iOS has been updated to `1.4.1-beta.1`. This version handles the `parentID` property during the RemotePost encoding, setting the dictionary value to `false` if the `parentID` is nil.

**Implementation:**
`Page` now contains 3 new properties that define if the page is a *top level page* and the hierarchy index
`Array+Page.swift` is the extension that contains all the functions to sort and set the properties to display the hierarchy view.
`PageListViewController` has been modified in order to display the ActionSheet menu button (if possible) and to detect some action from the parent view controller.
`PageListTableViewHandler` extends WPTableViewHandler. It's the new controller to handle different `NSFetchResultControllers` and the `TableView` datasource. It also sorts the pages in order to display the hierarchy.
`ParentPageSettingsViewController` is the new view controller used to display and select the available parent pages.

**To test:**
- From *Site Pages* section the `TableView` should have a *hierarchy* only when the filter *Published* is selected (and if at least one page has a parent page).
- Published, Draft and Trashed have a flat list, not grouped by date anymore.
- Scheduled keeps the grouped list by date.
- A *parent page* can be selected only from Published, Draft and Scheduled

danielebogo commented on WordPress/Classes/Extensions/Array+Page.swift:
```diff
@@ -0,0 +1,127 @@
+            right = dropFirst()
+
+        case lastIndex:
+            return Array<Element>(dropLast())
```
Shorthand for that case doesn't work

danielebogo commented:
I don't know why but Hound wrote some warning about different cases I never saw before:

- Optional closure `((_ page: AbstractPost?, _ error: Error?) -> Void)?)`
- Nested level for more than 1
- Shorthand syntactic sugar for cases where I can't use it like `return Array<Element>(dropLast())`

jklausa commented on WordPress/Classes/ViewRelated/Pages/PageListViewController.swift:
```diff
@@ -532,6 +573,34 @@ class PageListViewController: AbstractPostListViewController, UIViewControllerRe
+    private func addSetParentAction(to controller: UIAlertController, for page: AbstractPost, at index: IndexPath?) {
+        /// This button is disabled for self-hosted sites
+        //
+        guard blog.supports(.wpComRESTAPI) else {
```
This is a feature that core has too, right? Is the API different enough that it makes it difficult to implement the XML-RPC version, or is it just the case of handling WPCom API first?

danielebogo replied:
The second you said at the moment. I got and I still have problem with XMLRPC version. We actually read and store the value from the get posts but we don't update it. Also if you try to update that value the POST response of the UPDATE action is success but I still get the previous values from the get posts when the app sync again the pages list. I'm still investigating why.

jklausa commented on WordPress/Classes/ViewRelated/Pages/ParentPageSettingsViewController.swift:
```diff
@@ -0,0 +1,230 @@
+    // MARK: - Private methods
+
+    private func setupUI() {
+        WPFontManager.loadNotoFontFamily()
```
that shouldn't be needed here :) if the fonts haven't loaded by the time this VC gets on screen, we probably have much bigger problems :D

danielebogo replied:
I checked around and we use `WPFontManager.loadNotoFontFamily()` within the `viewDidLoad` (Aztec and Share Extension Editor). I also checked this class and apparently is safe. If it doesn't finish to load the font while it calls the `WPFontManager.notoRegularFont(ofSize: 17)`, the class first check if the font is loaded. If not it tries to load it again. If there's something wrong it fallbacks to the system font.
By the way because it's a factory class I can move it before the VC is created. What do you think?

jklausa replied:
>(Aztec and Share Extension Editor)

That makes sense though, because those can be launched outside of the context of the app (as you said, share extension!). This VC relies on the rest of the app having been set up, so the call to load fonts is superfluous here (and I think, actually close to a no-op given the implementation is wrapped in `dispatch_once`).

jklausa commented:
Ah, I forgot about to touch on one point: your `Row`/`Section` UITableView helpers are very similar in spirit to the `ImmuTable` we're using across the app — maybe it'd be worth looking into whether that'd be worth adopting here :)

danielebogo commented:
@jklausa have a look to my comment above  😊
Also I used the ImmutableRow and section as you suggested!
About this, _One other thing I noted while playing with it: I think it would be worthwhile (in a future follow-up PR) to change the "set parent" VC to also display the existing hierarchy and use the same font sizes/styles. When I was playing around with it, in big hierarchies (like Hogwarts P2) I had problem finding the page I wanted._, it's ok for me. I'd like to forward this to @SylvesterWilmott and see what he thinks about it 😃

jklausa commented:
>Also I used the ImmutableRow and section as you suggested!

✨  !

I think it doesn't need to be addressed/changed now, but you might want to look into `ImmuTableViewHandler` in the future too — it'd cut down on boilerplate UITableView code here considerably

jklausa commented:
oh wow those new hound warnings are some garbage, huh

SylvesterWilmott commented:
>change the "set parent" VC to also display the existing hierarchy and use the same font sizes/styles.

@danielebogo and I just discussed truncation, it seems the designs weren't clear but we should truncate at 1 line since in the vast majority of cases page titles will be short enough. In the event that we have a 2 line page title nested 10 levels deep that's going to turn into 4 lines. This is not much of an issue because we can press any page and open the editor to view the whole title.

The set parent page is slightly different, the cells should be dynamic because it's impossible to tap through to the editor here. If we did keep the hierarchy we would then need to truncate at 1 line as above.

I think the solution to what @jklausa mentioned here:

>When I was playing around with it, in big hierarchies (like Hogwarts P2) I had problem finding the page I wanted.

Is to include a search. @danielebogo how about a little scope creep?

danielebogo commented:
@SylvesterWilmott I will have to create another PR about the search anyway. Might worth include that too for a local search. In that VC all the loaded pages are injected into it.

danielebogo commented:
@SylvesterWilmott I fixed the Page Cell style, changing the title line number and the button width.
@jklausa A part from the non-self hosted filter and Hound's warnings, everything has been pushed and fixed! 😊

jklausa commented:
@danielebogo I still think the `WPFontManager.loadNotoFontFamily()` is not needed in the vincinity of those VC's at all — and I think I might be doing a bad job of explaining why, but otherwise :shipit:

danielebogo commented:
Thanks @jklausa !
